File: modules/shared/dart/subkey_service.dart
// BLAKE2b Sub-Key Derivation Service (Dart)
import 'dart:typed_data';
import 'dart:convert';
import 'package:sodium_libs/sodium_libs.dart';
import '../../features/security/services/sodium_instance.dart';

enum SubKeyContext { messages, notes, media }

class SubKeyService {
  static const String _contextString = 'MLINGO_1'; // exactly 8 ASCII bytes

  static const Map<SubKeyContext, int> _subKeyIds = {
    SubKeyContext.messages: 1,
    SubKeyContext.notes: 2,
    SubKeyContext.media: 3,
  };

  static Uint8List deriveSubKey({
    required Uint8List masterKey,
    required SubKeyContext context,
  }) {
    final int? subkeyId = _subKeyIds[context];
    if (subkeyId == null) {
      throw ArgumentError('CRYPTO_006: Unknown subkey context');
    }
    
    final derived = SodiumInstance.sodium.crypto.kdf.deriveFromKey(
      masterKey: SecureKey.fromList(SodiumInstance.sodium, masterKey),
      context: _contextString,
      subkeyId: BigInt.from(subkeyId),
      subkeyLen: 32,
    );
    return derived.extractBytes();
  }
}

File: modules/shared/node/subkey_service.js
// BLAKE2b Sub-Key Derivation Service (Node.js)
const sodium = require('libsodium-wrappers-sumo');

const CTX = 'MLINGO_1'; // exactly 8 ASCII bytes
const SUBKEY_IDS = {
  messages: 1,
  notes: 2,
  media: 3,
};

/**
 * Derives a 32-byte sub-key from a 32-byte master conversation key using crypto_kdf_derive_from_key.
 * @param {Buffer|Uint8Array} masterKey - 32 bytes
 * @param {'messages'|'notes'|'media'} context
 * @returns {Promise<Buffer>}
 */
async function deriveSubKey(masterKey, context) {
  await sodium.ready;
  const subkeyId = SUBKEY_IDS[context];
  if (!subkeyId) {
    throw new Error('CRYPTO_006: Unknown subkey context');
  }
  const subKey = sodium.crypto_kdf_derive_from_key(
    32,
    subkeyId,
    CTX,
    masterKey
  );
  return Buffer.from(subKey);
}

module.exports = {
  deriveSubKey,
};
